source/opt/constructs.h

```diff
@@ -0,0 +1,306 @@
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+// This file defines the language constructs for representing a SPIR-V
+// module in memory.
+
+#ifndef LIBSPIRV_OPT_CONSTRUCTS_H_
+#define LIBSPIRV_OPT_CONSTRUCTS_H_
+
+#include <functional>
+#include <vector>
+
+#include "spirv-tools/libspirv.h"
+#include "spirv/1.1/spirv.h"
+
+namespace spvtools {
+namespace opt {
+namespace ir {
+
+class Function;
+class Module;
+
+// About operand:
+//
+// In the SPIR-V specification, the term "operand" is used to mean any single
+// SPIR-V word following the leading wordcount-opcode word. Here, the term
+// "operand" is used to mean a *logical* operand. A logical operand may consist
+// of mulitple SPIR-V words, which together make up the same component. For
+// example, a logical operand of a 64-bit integer needs two words to express.
+//
+// Further, we categorize logical operands into *in* and *out* operands.
+// In operands are operands actually serve as input to operations, while out
+// operands are operands that represent ids generated from operations (result
+// type id or result id). For example, for "OpIAdd %rtype %rid %inop1 %inop2",
+// "%inop1" and "%inop2" are in operands, while "%rtype" and "%rid" are out
+// operands.
+
+// A *logical* operand to a SPIR-V instruction. It can be the type id, result
+// id, or other additional operands carried in an instruction.
+struct Operand {
+  Operand(spv_operand_type_t t, std::vector<uint32_t>&& w)
+      : type(t), words(std::move(w)) {}
+
+  spv_operand_type_t type;      // Type of this logical operand.
+  std::vector<uint32_t> words;  // Binary segments of this logical operand.
+
+  // TODO(antiagainst): create fields for literal number kind, width, etc.
+};
+
+// A SPIR-V instruction. It contains the opcode and any additional logical
+// operand. It may also contain line-related debug instruction (OpLine,
+// OpNoLine) directly appearing before this instruction.
+class Instruction {
+ public:
+  // Creates a default OpNop instruction.
+  Instruction() : opcode_(SpvOpNop), type_id_(0), result_id_(0) {}
+  // Creates an instruction with the given opcode |op| and no additional logical
+  // operands.
+  Instruction(SpvOp op) : opcode_(op), type_id_(0), result_id_(0) {}
+  // Creates an instruction using the given spv_parsed_instruction_t |inst|. All
+  // the data inside |inst| will be copied and owned in this instance. And keep
+  // record of line-related debug instructions |dbg_line| ahead of this
+  // instruction, if any.
+  Instruction(const spv_parsed_instruction_t& inst,
+              std::vector<Instruction>&& dbg_line = {});
+
+  SpvOp opcode() const { return opcode_; }
+  uint32_t type_id() const { return type_id_; }
+  uint32_t result_id() const { return result_id_; }
+  // Returns the vector of line-related debug instructions attached to this
+  // instruction and the caller can directly modify them.
+  std::vector<Instruction>& dbg_line_insts() { return dbg_line_insts_; }
```

Comment:
Do we want a const form of this accessor in case we want this data from a const Instruction?